### DISCUSSION TITLE
VectorDataReader

### DISCUSSION THREAD

**Discussion by Seddryck** [Giorgi/DuckDB.NET#149] [Ideas]
Hi @Giorgi 

Yesterday when trying to work on the PR, I was a bit struggling with the concept of VectorDataReader so I spent a few additional minutes to review this class. 

If my understanding is correct a vector in DuckDB is more or less a row and the class VectorDataReader allows us to read the value of a single field of this row. Is it correct (conceptually)?

**Comment by Giorgi** [comment]
No, VectorDataReader reads data for a specific column. The result object contains one or more chunks (depending on the number of rows in the result set) and each chunk has a vector for each column. If the column returns a composite type (List, Struct) the vector has child vector(s) that contain data. In the case of List, there is one child vector and in the case of Struct one vector for each property of struct.

**Reply by Seddryck** [reply]
Thx for clarification. 

I was wonderingn if this class doesn't deserve a refactoring. Most of the methods should throw exceptions as they are compatible with only a single type/value of the enum passed in parameter. It also leads to many nullable class properties that are available/needed in a specific context (decimal, list, ...). 

I would suggest to imply an abstract factory instantiating specific tiny classes just handling a specific column type. It should improve readability and maintainability. Do you want that I try to submit a PR?

**Reply by Giorgi** [reply]
I think that's a good idea but I haven't thought about it and don't know how easy or doable it is. If you want to try it please use the [Struct-Support](https://github.com/Giorgi/DuckDB.NET/tree/Struct-Support) as a base branch.

**Reply by Giorgi** [reply]
@Seddryck Still plan to implement it?

**Reply by Seddryck** [reply]
Yes will do it on Sunday.

**Reply by Giorgi** [reply]
I haven't thought much about it but perhaps instead of breaking it into many small classes it makes sense to break it into FlatVectorReader, ListVectorReader and StructDataReader?

**Reply by Seddryck** [reply]
At this moment I've not much more classes (but I'm still not validating with tests).

**Reply by Seddryck** [reply]
Current status: 
- worked on that 2-3 hours and still have the `struct` to go
- all tests passing (except `struct`)
- List as been a challenge due to list of nullable but solved
- one test sounds really weird. `ReadMultipleListOfDoubles` returns int (0,1,2) in place of (0.5, 1.5, 2.5). I don't think the issue is on my side as the FieldType returned is DuckDBInteger. Will check a little bit more but updated the test to force double returned by query.

**Reply by Giorgi** [reply]
You are probably using an older version of DuckDB that does integer division when using /

https://duckdb.org/docs/sql/functions/numeric.html

**Reply by Giorgi** [reply]
If you delete all bin/obj folders in the solution and rebuild the Test project, it will download the latest version of DuckDB.

**Reply by Seddryck** [reply]
Thx, it fixed the issue with the test.

Implemented for Struct and also added Blob which was missing in the factory. Initial implementation id done.
Feel free to review in the PR #153

**Reply by Giorgi** [reply]
What was missing for blobs?

**Reply by Seddryck** [reply]
Was my bad but didn't see it before because tests were on the namespace of parameters.

**Comment by Seddryck** [comment]
I've an issue with the consistency regarding the ClrTypes.

```csharp
DuckDBType.Timestamp => typeof(DateTime),
DuckDBType.Interval => typeof(DuckDBInterval),
DuckDBType.Date => typeof(DuckDBDateOnly),
DuckDBType.Time => typeof(DuckDBTimeOnly),
DuckDBType.HugeInt => typeof(BigInteger),
```

I'm not sure why we are returning a `BigInteger` and not `DuckDBHugeInt` but `DuckDBDateOnly` and not `DateOnly`. I guess it's linked to .NET Standard 2.0 vs .NET 6.0 but IMO, it's a bit inconsistent.

**Reply by Giorgi** [reply]
Yeah, I agree. It's not consistent. I'll need some time to analyze what effect changing it will have.

**Reply by Seddryck** [reply]
Don't put too much effort in the analysis, we'll see that later. Could be easier to evaluate after refactoring. Keeping the inconsistency as-is at the moment.

**Reply by Seddryck** [reply]
`DuckDBType.Enum => typeof(string),` sounds also weird.

**Reply by Giorgi** [reply]
It's because `GetValue` returns a string for Enum columns.

**Comment by Seddryck** [comment]
A bit perplex about this case:

The original code in `DuckDBDataReader` was
```csharp
public override bool GetBoolean(int ordinal)
{
    return GetByte(ordinal) != 0;
}
```
after refactoring, to pass the tests, I had to switch back to 
```csharp
public override bool GetBoolean(int ordinal)
{
    return GetFieldData<bool>(ordinal);
}
```
There was any reason to parse the boolean as a byte before? Am I missing something? (If yes we should change the tests because it's not capturing something).

**Reply by Seddryck** [reply]
Thinking out that I could read only the first bit of the byte and by luck getting the expected result. Will also check on my side.

**Reply by Giorgi** [reply]
There is no specific reason why I wrote it that way, it was the first thing that came to my mind. But I'm wondering if you are only refactoring and moving the code around, why would it cause that to fail. Either there was already a bug in `GetBoolean` that you uncovered now (which is completely possible) or you are doing something wrong.

**Comment by Seddryck** [comment]
I'm blocked somewhere with `object GetEnum(ulong offset, Type returnType)` which can return `null` at `(line 239) return default!;` Which means at the end that `T GetFieldValue<T>` in `DuckDBDataReader` can also return `null` which is not aligned with the `SqlClient.SqlDataReader` throwing an exception when it hits a `null` value. Also `T GetFieldValue<T>` is not supposed to return `null` or it would be T? but this method overrides a method defined in DbDataReader so we can't really changing this behavior.

That wouldn't be a big problem to change this behavior but there is a test specifically testing this behavior (SelectEnumsAsNullable). 

What do we do? Back to the standard behavior or continue to cheat with the compiler?

**Reply by Seddryck** [reply]
I checked in Npgsql is also returning `null` in [case](https://github.com/npgsql/npgsql/blob/4ccfcd71b690afd4f8e2aa4e7c6642826a92d356/src/Npgsql/NpgsqlDataReader.cs#L2106..2108) of `Nullable<>`.

**Reply by Giorgi** [reply]
`GetFieldValue` will return null only for Nullable types and if the value is null. You are right about Postgres, I asked npgsql maintainer for an advice.
